### PR TITLE
FEATURE: LX-23 Introduce available value filter

### PR DESCRIPTION
**Pull request by xMort**
JIRA: LX-23

---

Supported PR commands:

| Command                                                 | Description                                                |
| ------------------------------------------------------- | ---------------------------------------------------------- |
| `ok to test`                                            | Re-run standard checks                                     |
| `extended check sonar`                                  | SonarQube tests                                            |
| `extended test - backstop`                              | BackstopJS tests                                           |
| **E2E Cypress tests commands - TIGER**                  |                                                            |
| `extended test - tiger-cypress - isolated <testName>`   | Run isolated tests running against recorded Tiger backend. |
| `extended test - tiger-cypress - record <testName>`     | Create a new recording for isolated Tiger tests.           |
| `extended test - tiger-cypress - integrated <testName>` | Run integrated tests against live backend                  |
| **E2E Cypress tests commands - BEAR**                   |                                                            |
| `extended test - cypress - isolated <testName>`         | Run isolated tests running against recorded Bear backend.  |
| `extended test - cypress - record <testName>`           | Create a new recording for isolated Bear tests.            |
| `extended test - cypress - integrated <testName>`       | Run integrated tests against live backend                  |
| **Compatibility matrix test commands - TIGER Backend**  |                                                            |
| `extended test - matrix-test <AIO_version>`             | Run integrated tests against AIO versions.                 |

`<testName>` in cypress commands is used to filter specfiles. Example, to run record with BEAR backend

-   Against `dashboard.spec.ts` and `drilling.spec.ts`, execute command `extended test - cypress - record dashboard,drilling`
-   Against all specfiles, execute command `extended test - cypress - record` or `extended test - cypress - record *`

`<AIO_version>` in commands is used to start test with multiple AIO instances - each instance in triggered by one jenkins build

-   To run with `master` and `stable`, execute command `extended test - matrix-test master,stable` or `extended test - matrix-test latest`
-   To run with specific version,ex: `2.3.0` and `2.3.1`, execute command `extended test - matrix-test 2.3.0,2.3.1`
-   In case `<AIO_version>` is empty, read versions from file `compTigerVersions.txt` of this repo

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `extended test - backstop` passes
-   [ ] `extended test - tiger-cypress - record` to record new mapping files (Tiger BE)
-   [ ] `extended test - cypress - record` to record new mapping files (Bear BE)
-   [ ] `extended test - tiger-cypress - isolated` passes
-   [ ] `extended test - cypress - isolated` passes
-   [ ] `extended test - tiger-cypress - integrated` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17427/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18938/ : FAILURE in 2m 03s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17429/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18943/ : FAILURE in 11m 55s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17441/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18960/ : SUCCESS in 13m 37s

File: libs/sdk-model/src/dashboard/filterContext.ts [hunk 1]
```diff
@@ -47,6 +47,18 @@ export interface IDashboardAttributeFilterParent {
...
+/**
+ * Item that is used to validate availability of attribute element.
+ *
+ * The item is either metric or is converted to metric (in case of attribute, label, or fact) by backend
+ * and cross-tabulation report is executed to determine if attribute element should be available (visible).
+ *
+ * @alpha
+ */
+export interface IElementValidationItem {
+    identifier: ObjRef;
```
Comment (ivan-nejezchleb) [inline libs/sdk-model/src/dashboard/filterContext.ts:59]: would not be ref, itemRef or something similar better? "identifier" implies limited usage to idRef, which is probably true, but should not leak to the model level

File: libs/sdk-model/src/dashboard/filterContext.ts [hunk 2]
```diff
@@ -47,6 +47,18 @@ export interface IDashboardAttributeFilterParent {
  */
 export type DashboardAttributeFilterSelectionMode = "single" | "multi";
 
+/**
+ * Item that is used to validate availability of attribute element.
+ *
+ * The item is either metric or is converted to metric (in case of attribute, label, or fact) by backend
```
Comment (ivan-nejezchleb) [inline libs/sdk-model/src/dashboard/filterContext.ts:53]: can we comment how it is converted, which aggregation functions are used? Just to describe behavior to reader. Thx

Comment (xMort) [reply libs/sdk-model/src/dashboard/filterContext.ts:53]: Fixed :)

File: libs/sdk-model/src/dashboard/filterContext.ts (hunk 1 above)
Comment (ivan-nejezchleb) [reply libs/sdk-model/src/dashboard/filterContext.ts:60]: ok, thx for explanation. I had no idea about this automagic

**Review by ivan-nejezchleb — APPROVED** [review]

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17470/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18988/ : SUCCESS in 11m 41s

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR4509

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/3708/ : SUCCESS in 8m 38s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18998/ : SUCCESS in 12m 14s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-bear-isolated-zuul-pipeline/1336/ : UNSTABLE in 34m 27s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/4876/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/1587/ : FAILURE in 2m 16s (non-voting)
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-tiger-isolated-zuul-pipeline/1289/ : SUCCESS in 14m 14s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-compatibility-pipeline/577/ : SUCCESS in 6m 36s

**Comment by xMort** [comment]
extended test - cypress - isolated

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17580/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/1885/ : UNSTABLE in 43m 26s

**Comment by xMort** [comment]
extended test - cypress - isolated

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17584/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/1886/ : UNSTABLE in 11m 23s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/17583/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/19108/ : SUCCESS in 11m 54s

**Comment by xMort** [comment]
👆 Check extended pipeline failed due to drilling issue. Seems flaky as I did not touch that behavior.

**Comment by tuan-tran-thanh** [comment]
@xMort , yes that is the flaky test, I will fix it

File: libs/sdk-backend-tiger/src/backend/workspace/attributes/elements/index.ts [hunk 3]
```diff
@@ -102,7 +106,14 @@ class TigerWorkspaceElementsQuery implements IElementsQuery {
     }
 
     public withMeasures(): IElementsQuery {
-        throw new NotSupported("withMeasures is not supported in sdk-backend-tiger yet");
+        throw new NotSupported(
+            "withMeasures is not supported in sdk-backend-tiger yet. Use withValidateElementsOnly instead.",
```
[the remark this answers is not in the document]
Comment (xMort) [reply libs/sdk-backend-tiger/src/backend/workspace/attributes/elements/index.ts:110]: Fixed.

File: libs/sdk-ui-filters/api/sdk-ui-filters.api.md [hunk 4]
```diff
@@ -796,6 +800,8 @@ export interface ILoadElementsOptions {
     // (undocumented)
     limitingMeasures?: IMeasure[];
     // (undocumented)
+    limitingValidationItems?: ObjRef[];
```
[the remark this answers is not in the document]
Comment (xMort) [reply libs/sdk-ui-filters/api/sdk-ui-filters.api.md:803]: New capability created and used in loadElementsFromBackend.

**Review by ivan-nejezchleb — APPROVED** [review]

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR4509

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/3755/ : SUCCESS in 8m 23s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/19114/ : SUCCESS in 11m 40s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-bear-isolated-zuul-pipeline/1386/ : SUCCESS in 9m 38s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/4923/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/1634/ : FAILURE in 1m 15s (non-voting)
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-tiger-isolated-zuul-pipeline/1375/ : SUCCESS in 14m 53s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-compatibility-pipeline/579/ : SUCCESS in 5m 52s

**Pull request merged**